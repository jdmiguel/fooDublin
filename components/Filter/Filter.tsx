import React, { useReducer, Dispatch } from 'react';
import styled from 'styled-components';

type FilterData = {
  text: string;
  sort: 'cost' | 'rating';
  order: 'asc' | 'desc';
  id: number;
};

type FilterDataItemTypeWithIsActive = FilterData & { isActive: boolean };

type FilterProps = {
  className?: string;
  onSelect: (sort: string, order: string) => void;
  data: FilterData[];
};

type DataAction = { type: 'select'; id: number } | { type: 'clear' };

const StyledFilter = styled.div`
  width: 100%;
  max-width: 71rem;
  padding: 20px 0 10px;
  display: flex;
  justify-content: center;
  flex-wrap: wrap;
  background-color: ${(props) => props.theme.palette.LIGHT_MAX};
  border-bottom: 1px solid ${(props) => props.theme.palette.LIGHT_SOFT};
  @media only screen and (min-width: 480px) {
    padding: 20px 20px 10px;
  }
  @media only screen and (min-width: 768px) {
    padding: 25px 0;
    justify-content: space-evenly;
  }
  @media only screen and (min-width: 992px) {
    border: 1px solid ${(props) => props.theme.palette.LIGHT_SOFT};
    border-radius: 4px;
  }
`;

const StyledFilterItem = styled.button<{ isActive: boolean }>`
  width: 144px;
  padding: 15px 8px;
  margin: 0 8px 15px;
  border: 1px solid ${(props) => props.theme.palette.DARK_MIN};
  background-color: ${(props) => props.theme.palette.LIGHT_MEDIUM};
  font-size: 0.9rem;
  color: ${(props) => props.theme.palette.DARK_MEDIUM};
  outline: none;
  cursor: pointer;
  background-color: ${({ theme, isActive }) =>
    `${isActive ? theme.palette.PRIMARY_LIGHT : theme.palette.LIGHT_MEDIUM}`};
  @media only screen and (min-width: 400px) {
    width: 180px;
  }
  @media only screen and (min-width: 480px) {
    width: 200px;
    margin: 0 10px 15px;
  }
  @media only screen and (min-width: 680px) {
    width: 250px;
  }
  @media only screen and (min-width: 768px) {
    width: auto;
    margin: 0;
    padding: 15px;
    font-size: 1rem;
  }
  @media only screen and (min-width: 992px) {
    padding: 20px;
  }
`;

const dataReducer = (data: FilterData[], action: DataAction) => {
  switch (action.type) {
    case 'select':
      return data.map((dataItem) => ({
        ...dataItem,
        isActive: action.id === dataItem.id,
      }));
    case 'clear':
      return data.map((dataItem) => ({
        ...dataItem,
        isActive: false,
      }));
  }
};

const Filter: React.FC<FilterProps> = ({ className, onSelect, data }) => {
  const dataWithIsActiveProp = data.map((dataItem) => ({
    ...dataItem,
    isActive: false,
  }));
  const [dataState, dispatch]: [
    FilterDataItemTypeWithIsActive[],
    Dispatch<DataAction>,
  ] = useReducer(dataReducer, dataWithIsActiveProp);

  const handleSelect = (id: number, sort: string, order: string) => {
    dispatch({ type: 'select', id });
    onSelect && onSelect(sort, order);
  };

  return (
    <StyledFilter className={className}>
      {dataState.map((item: FilterDataItemTypeWithIsActive) => (
        <StyledFilterItem
          isActive={item.isActive}
          key={item.id}
          onClick={(event: React.MouseEvent<HTMLElement>) => {
            event.preventDefault();
            handleSelect(item.id, item.sort, item.order);
          }}
        >
          {item.text}
        </StyledFilterItem>
      ))}
    </StyledFilter>
  );
};

export default Filter;
