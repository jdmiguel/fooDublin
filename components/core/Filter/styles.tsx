import styled from 'styled-components';

export const StyledFilter = styled.button<{ isActive: boolean }>`
  padding: 15px 8px;
  font-size: 0.9rem;
  border: 1px solid ${({ theme }) => theme.palette.LIGHT_SOFT};
  border-radius: 4px;
  color: ${({ theme }) => theme.palette.PRIMARY_MEDIUM};
  outline: none;
  cursor: pointer;
  display: flex;
  justify-content: center;
  align-items: center;
  flex: 1;
  transition: all 0.2s ease-out;
  background-color: ${({ theme, isActive }) =>
    `${isActive ? theme.palette.PRIMARY_LIGHT : theme.palette.LIGHT_MEDIUM}`};
  @media only screen and (min-width: 768px) {
    width: auto;
    margin: 0;
    padding: 15px;
    font-size: 1rem;
  }
  @media only screen and (min-width: 1200px) {
    &:hover {
      background-color: ${({ theme }) => theme.palette.PRIMARY_LIGHT};
      color: ${({ theme }) => theme.palette.PRIMARY};
    }
  }
`;

export const StyledFilterText = styled.span`
  text-transform: uppercase;
  font-weight: 500;
`;

export const StyledFilterIcon = styled.i`
  font-size: 1.1rem;
  margin-right: 8px;
`;
