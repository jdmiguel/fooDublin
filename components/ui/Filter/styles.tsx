import styled from 'styled-components';

export const StyledFilterItem = styled.button<{ isActive: boolean }>`
  width: 144px;
  padding: 15px 8px;
  margin: 0 8px 15px;
  border: 1px solid ${({ theme }) => theme.palette.DARK_MIN};
  font-size: 0.9rem;
  color: ${({ theme }) => theme.palette.DARK_MEDIUM};
  outline: none;
  cursor: pointer;
  display: flex;
  justify-content: center;
  transition: background-color 0.2s ease-out;
  background-color: ${({ theme, isActive }) =>
    `${isActive ? theme.palette.PRIMARY_LIGHT : theme.palette.LIGHT_MIN}`};
  &:hover {
    background-color: ${({ theme }) => theme.palette.PRIMARY_LIGHT};
  }
  @media only screen and (min-width: 400px) {
    width: 180px;
  }
  @media only screen and (min-width: 480px) {
    width: 200px;
    margin: 0 10px 15px;
  }
  @media only screen and (min-width: 680px) {
    width: 250px;
  }
  @media only screen and (min-width: 768px) {
    width: auto;
    margin: 0;
    padding: 15px;
    font-size: 1rem;
  }
  @media only screen and (min-width: 992px) {
    padding: 20px;
  }
`;

export const StyledFilterPrimaryText = styled.span`
  text-transform: uppercase;
`;

export const StyledFilterSecondaryText = styled.span`
  display: none;
  @media only screen and (min-width: 920px) {
    display: block;
  }
`;

export const StyledFilterIcon = styled.i`
  font-size: 1.1rem;
  line-height: 18px;
  margin-left: 10px;
  display: block;
  @media only screen and (min-width: 920px) {
    display: none;
  }
`;
