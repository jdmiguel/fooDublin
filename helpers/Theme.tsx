import React from 'react';
import { DefaultTheme, ThemeProvider } from 'styled-components';

export const theme: DefaultTheme = {
  palette: {
    PRIMARY: '#EC7E00',
    PRIMARY_DARK: '#D48C24',
    PRIMARY_MEDIUM: '#FBA52B',
    PRIMARY_LIGHT: '#FDF3E5',
    SECONDARY: '#660000',
    LIGHT_MAX: '#FFFFFF',
    LIGHT_MEDIUM: '#F3F3F3',
    LIGHT_MIN: '#E5EAED',
    LIGHT_SOFT: '#D8D8D8',
    DARK_MAX: '#222C3C',
    DARK_MEDIUM: '#585858',
    DARK_MIN: '#89959B',
    DARK_SOFT: '#A7A7A7',
  },
};

export const renderWithTheme = (component: any) => (
  <ThemeProvider theme={theme}>{component}</ThemeProvider>
);
