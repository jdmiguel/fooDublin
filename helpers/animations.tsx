import { keyframes, css } from 'styled-components';

const fadeIn = (endValue: number) => keyframes`
  0% { opacity: 0 }
  100% { opacity: ${endValue} }
`;

const fadeInFromPosY = (initialPosY: number) => keyframes`
  0% { opacity: 0; transform: translateY(${initialPosY}px); }
  100% { opacity: 1; ; transform: translateY(0); }
`;

const rotate = keyframes`
  0% { transform: rotate(0deg); }
  100% { transform: rotate(360deg); }
`;

const moveX = keyframes`
  0% { transform: translateX(-100%); }
  100% { transform: translateX(0); }
`;

export const fadeInAnimation = css<{ endValue?: number }>`
  animation: ${({ endValue }) => fadeIn(endValue || 1)} 0.15s linear;
`;

export const fadeInFromPosYAnimation = css<{
  initialPosY?: number;
  animationDuration?: number;
  animationDelay?: number;
}>`
  animation-name: ${({ initialPosY }) => fadeInFromPosY(initialPosY || 50)};
  animation-duration: ${({ animationDuration }) => animationDuration || 0.15}s;
  animation-delay: ${({ animationDelay }) => animationDelay || 0.25}s;
  animation-fill-mode: forwards;
  animation-timing-function: ease-out;
`;

export const circleLoaderAnimation = css`
  animation: ${rotate} 0.8s linear infinite;
`;

export const lineLoaderAnimation = css`
  animation: ${moveX} 1s ease-in-out infinite forwards;
`;
