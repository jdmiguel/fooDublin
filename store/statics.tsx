import { BreadcrumbsType } from '@/components/core/types';

export const BASE_API = 'https://api.yelp.com/v3/';

export const DUBLIN_ID = 91;

export const DEFAULT_TEXT_LOADING = 'Coming right up...';

export const SEARCH_TERM_QUERY_PARAM = 'restaurants';
export const SEARCH_RADIUS_QUERY_PARAM = 1000;

export const MAX_RESTAURANT_DISPLAYED = 30;

export const MAX_SMALL_DEVICE_WIDTH = 640;

export const SCROLL_FACTOR = 1.1;
export const SCROLL_DELAY = 100;

export const SHOWING_SCROLLUP_BUTTON_HEIGHT = 900;

export const MAX_SEARCH_BREADCRUMBS = 2;

export const MAX_MOBILE_WIDTH = 768;

export enum PlaceholderText {
  BLURRED = 'Search for locals...',
  FOCUSED = 'Start typing to search...',
}

export const HIGHLIGHT_GENERIC_SRC = '/images/highlight-generic.png';
export const THUMB_GENERIC_SRC = '/images/generic-thumb.png';
export const DETAIL_GENERIC_SRC = '/images/generic-detail.png';
export const USER_GENERIC_SRC = '/images/user-avatar.svg';
export const DEFAULT_RATING_STAR_LIST = [
  { id: 1, type: 'star_outline' },
  { id: 2, type: 'star_outline' },
  { id: 3, type: 'star_outline' },
  { id: 4, type: 'star_outline' },
  { id: 5, type: 'star_outline' },
];

export enum FilterSort {
  COST = 'cost',
  RANK = 'rating',
}
export enum FilterOrder {
  ASC = 'asc',
  DESC = 'desc',
}

export const FILTERS = [
  {
    isActive: false,
    primaryText: 'cost',
    secondaryText: '- high to low',
    icon: 'keyboard_arrow_up',
    sort: FilterSort.COST,
    order: FilterOrder.ASC,
    id: 1,
  },
  {
    isActive: false,
    primaryText: 'cost',
    secondaryText: '- low to high',
    icon: 'keyboard_arrow_down',
    sort: FilterSort.COST,
    order: FilterOrder.DESC,
    id: 2,
  },
  {
    isActive: false,
    primaryText: 'rank',
    secondaryText: '- high to low',
    icon: 'keyboard_arrow_up',
    sort: FilterSort.RANK,
    order: FilterOrder.ASC,
    id: 3,
  },
  {
    isActive: false,
    primaryText: 'rank',
    secondaryText: '- low to high',
    icon: 'keyboard_arrow_down',
    sort: FilterSort.RANK,
    order: FilterOrder.DESC,
    id: 4,
  },
];

export const DEFAULT_BREADCRUMB = {
  text: 'Home',
  route: '/',
  asRoute: '/',
  type: BreadcrumbsType.HOME,
};

export const DUBLIN_COORDINATES = {
  latitude: 53.3434,
  longitude: -6.26761,
};

export const WEEK_DAYS = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'];

export const MAX_PRICE_PERCENT = 80;
export const MAX_EURO_PRICE_AMOUNT = 4;
