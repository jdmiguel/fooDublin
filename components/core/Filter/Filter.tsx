import { StyledFilter, StyledFilterText, StyledFilterIcon } from './styles';
import { FilterData, FilterType } from '../types';

type FilterProps = {
  className?: string;
  data: FilterData;
  onClick: (filter: FilterType) => void;
};

export const Filter: React.FC<FilterProps> = ({
  data: { isActive, text, icon, type },
  onClick,
}) => (
  <StyledFilter isActive={isActive} onClick={() => onClick(type)}>
    <StyledFilterText>{text}</StyledFilterText>
    <StyledFilterIcon className="material-icons">{icon}</StyledFilterIcon>
  </StyledFilter>
);
