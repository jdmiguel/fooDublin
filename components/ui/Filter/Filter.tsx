import React, { useReducer, Dispatch } from 'react';
import styled from 'styled-components';

import { FilterSort, FilterOrder } from '../../../helpers/staticData';

type FilterData = {
  primaryText: string;
  secondaryText: string;
  icon: string;
  sort: FilterSort.COST | FilterSort.RANK;
  order: FilterOrder.ASC | FilterOrder.DESC;
  id: number;
};

type FilterDataItemTypeWithIsActive = FilterData & { isActive: boolean };

type FilterProps = {
  className?: string;
  onClick: (sort: string, order: string) => void;
  data: FilterData[];
};

type DataAction = { type: 'select'; id: number } | { type: 'clear' };

const StyledFilterItem = styled.button<{ isActive: boolean }>`
  width: 144px;
  padding: 15px 8px;
  margin: 0 8px 15px;
  border: 1px solid ${({ theme }) => theme.palette.DARK_MIN};
  font-size: 0.9rem;
  color: ${({ theme }) => theme.palette.DARK_MEDIUM};
  outline: none;
  cursor: pointer;
  display: flex;
  justify-content: center;
  transition: background-color 0.2s ease-out;
  background-color: ${({ theme, isActive }) =>
    `${isActive ? theme.palette.PRIMARY_LIGHT : theme.palette.LIGHT_MIN}`};
  &:hover {
    background-color: ${({ theme }) => theme.palette.PRIMARY_LIGHT};
  }
  @media only screen and (min-width: 400px) {
    width: 180px;
  }
  @media only screen and (min-width: 480px) {
    width: 200px;
    margin: 0 10px 15px;
  }
  @media only screen and (min-width: 680px) {
    width: 250px;
  }
  @media only screen and (min-width: 768px) {
    width: auto;
    margin: 0;
    padding: 15px;
    font-size: 1rem;
  }
  @media only screen and (min-width: 992px) {
    padding: 20px;
  }
`;

const StyledFilterPrimaryText = styled.span`
  text-transform: uppercase;
`;

const StyledFilterSecondaryText = styled.span`
  display: none;
  @media only screen and (min-width: 920px) {
    display: block;
  }
`;

const StyledFilterIcon = styled.i`
  font-size: 1.1rem;
  line-height: 18px;
  margin-left: 10px;
  display: block;
  @media only screen and (min-width: 920px) {
    display: none;
  }
`;

const dataReducer = (data: FilterData[], action: DataAction) => {
  switch (action.type) {
    case 'select':
      return data.map((dataItem) => ({
        ...dataItem,
        isActive: action.id === dataItem.id,
      }));
    case 'clear':
      return data.map((dataItem) => ({
        ...dataItem,
        isActive: false,
      }));
  }
};

const Filter: React.FC<FilterProps> = ({ onClick, data }) => {
  const dataWithIsActiveProp = data.map((dataItem) => ({
    ...dataItem,
    isActive: false,
  }));
  const [dataState, dispatch]: [
    FilterDataItemTypeWithIsActive[],
    Dispatch<DataAction>,
  ] = useReducer(dataReducer, dataWithIsActiveProp);

  const handleClick = (
    isActive: boolean,
    id: number,
    sort: string,
    order: string,
  ) => {
    dispatch(!isActive ? { type: 'select', id } : { type: 'clear' });
    onClick(!isActive ? sort : '', !isActive ? order : '');
  };

  return (
    <div className="grid-x grid-margin-x">
      {dataState.map((item) => (
        <StyledFilterItem
          isActive={item.isActive}
          className="cell small-6 medium-3 large-3"
          key={item.id}
          onClick={() => {
            handleClick(item.isActive, item.id, item.sort, item.order);
          }}
        >
          <StyledFilterPrimaryText>{item.primaryText}</StyledFilterPrimaryText>
          <StyledFilterSecondaryText>
            {` ${item.secondaryText}`}
          </StyledFilterSecondaryText>
          <StyledFilterIcon className="material-icons">
            {item.icon}
          </StyledFilterIcon>
        </StyledFilterItem>
      ))}
    </div>
  );
};

export default Filter;
