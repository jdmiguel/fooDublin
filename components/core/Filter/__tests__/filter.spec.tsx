/**
 * @jest-environment jsdom
 */

import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import 'jest-styled-components';
import { Filter } from '../Filter';
import { renderWithTheme } from '../../../../helpers/Theme';
import { FILTERS } from '../../../../store/statics';

describe('Component: Filter', () => {
  const [firstFilter] = FILTERS;

  const props = {
    onClick: jest.fn(),
  };

  it('should render correctly', () => {
    const { container } = render(renderWithTheme(<Filter {...props} data={firstFilter} />));

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should call callback function on click', async () => {
    const handleClick = jest.fn();

    render(renderWithTheme(<Filter onClick={handleClick} data={firstFilter} />));

    await userEvent.click(screen.getByRole('button'));

    expect(handleClick).toHaveBeenCalled();
  });
});
