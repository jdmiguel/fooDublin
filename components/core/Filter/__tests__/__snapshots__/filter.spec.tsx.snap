// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Filter should render correctly 1`] = `
.c0 {
  padding: 15px 8px;
  font-size: 0.9rem;
  border: 1px solid #D8D8D8;
  border-radius: 4px;
  color: #FBA52B;
  outline: none;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-transition: all 0.2s ease-out;
  transition: all 0.2s ease-out;
  background-color: #F3F3F3;
}

.c2 {
  text-transform: uppercase;
  font-weight: 500;
}

.c1 {
  font-size: 1.1rem;
  margin-right: 8px;
}

@media only screen and (min-width:768px) {
  .c0 {
    width: auto;
    margin: 0;
    padding: 15px;
    font-size: 1rem;
  }
}

@media only screen and (min-width:1200px) {
  .c0:hover {
    background-color: #FDF3E5;
    color: #EC7E00;
  }
}

<button
  class="c0"
>
  <i
    class="c1 material-icons"
  >
    monetization_on
  </i>
  <span
    class="c2"
  >
    high cost
  </span>
</button>
`;
